### PR TITLE
[Eager] Support test imperative basic in eager  test_empty_grad

### PR DESCRIPTION
**Pull request by wanghuancoder**
### PR types

New features
### PR changes

Others 
### Describe

- 调通了单测 test_empty_grad
- eager_tensor_method_copy_，只在!self->eager_tensor.defined()时，才设置StopGradient和Persistable
- 为eagertensor添加了clear_gradient函数
- 重新实现了EagerTensor的Patch逻辑

### CODE REVIEW

File: paddle/fluid/pybind/eager_method.cc [hunk 1]
```diff
@@ -19,6 +19,7 @@ limitations under the License. */
 
 #include "paddle/fluid/eager/accumulation/accumulation_node.h"
 #include "paddle/fluid/eager/api/all.h"
+#include "paddle/fluid/eager/api/generated/fluid_generated/dygraph_forward_api.h"
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_method.cc]: Why this?

File: paddle/fluid/pybind/eager_method.cc [hunk 2]
```diff
@@ -152,6 +157,61 @@ static PyObject* eager_tensor_retain_grads(EagerTensorObject* self,
...
+          (int)pten::TransToProtoVarType(dense_tensor->dtype());  // NOLINT
+      auto ddim = grad.shape();
+      std::vector<int64_t> value;
+      size_t rank = static_cast<size_t>(ddim.size());
+      value.resize(rank);
+      for (size_t i = 0; i < rank; i++) {
+        value[i] = ddim[i];
+      }
+      attrs["shape"] = value;
+      fill_constant_dygraph_function(&grad, attrs);
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_method.cc]: no need to use `fill_constant_dygraph_function`, just use `paddle::experimental::ones_like` instead

File: paddle/fluid/pybind/eager_method.cc [hunk 3]
```diff
@@ -152,6 +157,61 @@ static PyObject* eager_tensor_retain_grads(EagerTensorObject* self,
...
+    grad = accumulation_grad_node->Grad();
+  } else {
+    auto meta = egr::EagerUtils::unsafe_autograd_meta(self->eager_tensor);
+    grad = meta->Grad();
+  }
+
+  if (grad.initialized()) {
+    auto dense_tensor =
+        std::dynamic_pointer_cast<pten::DenseTensor>(grad.impl());
+    if (set_to_zero) {
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_method.cc]: remove this parameter and add one more API like `zero_grads`

File: python/paddle/__init__.py [hunk 4]
```diff
@@ -24,8 +24,13 @@
 from .batch import batch  # noqa: F401
 from .fluid import monkey_patch_variable
 from .fluid.dygraph import monkey_patch_math_varbase
+from .fluid import core
+from paddle.fluid.framework import _test_eager_guard
```
Comment (JiabinYang) [inline python/paddle/__init__.py]: Add this to `__init__` may cause used end can see it, find another way maybe

Comment (JiabinYang) [reply python/paddle/__init__.py]: How about patch it in `_test_eager_guard` when we use `_test_eager_guard`

File: python/paddle/fluid/__init__.py [hunk 5]
```diff
@@ -220,7 +217,9 @@ def remove_flag_if_exists(name):
 monkey_patch_variable()
 __bootstrap__()
 monkey_patch_varbase()
-monkey_patch_eagertensor()
+if hasattr(core, "eager"):
```
Comment (JiabinYang) [inline python/paddle/fluid/__init__.py]: same above

File: python/paddle/fluid/initializer.py [hunk 6]
```diff
@@ -604,7 +604,7 @@ def __call__(self, var, block=None):
         if framework.in_dygraph_mode():
             if self._uniform:
                 limit = np.sqrt(6.0 / float(fan_in + fan_out))
-                out_var = _C_ops.uniform_random('shape', var.shape, 'min',
+                out_var = _C_ops.uniform_random('shape', out_var.shape, 'min',
```
Comment (JiabinYang) [inline python/paddle/fluid/initializer.py:607]: why this?

**Review by JiabinYang — COMMENTED** [review]
some comment

File: paddle/fluid/pybind/eager_method.cc (hunk 1 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_method.cc]: 是因为使用了fill_constant_dygraph_function，已经删掉了

File: paddle/fluid/pybind/eager_method.cc (hunk 2 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_method.cc]: done，thx！

File: python/paddle/fluid/initializer.py (hunk 6 above)
Comment (wanghuancoder) [reply python/paddle/fluid/initializer.py:607]: 这是为了与静态图逻辑保持一致。代码下方，静态图append_op，传入的就是out_var.shape。

File: python/paddle/__init__.py (hunk 4 above)
Comment (wanghuancoder) [reply python/paddle/__init__.py]: done，thx！

File: python/paddle/fluid/__init__.py (hunk 5 above)
Comment (wanghuancoder) [reply python/paddle/fluid/__init__.py]: done，thx！

File: paddle/fluid/pybind/eager_method.cc (hunk 3 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_method.cc]: done, thx!

File: paddle/fluid/pybind/eager_method.cc [hunk 7]
```diff
@@ -152,6 +157,72 @@ static PyObject* eager_tensor_retain_grads(EagerTensorObject* self,
...
+    auto dense_tensor =
+        std::dynamic_pointer_cast<pten::DenseTensor>(grad.impl());
+    dense_tensor->release();
+  }
+  Py_INCREF(Py_None);
+  return Py_None;
+  EAGER_CATCH_AND_THROW_RETURN_NULL
+}
+
+static PyObject* eager_tensor_zero_grads(EagerTensorObject* self,
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_method.cc]: make this method `eager_tensor__zero_grads`

File: paddle/fluid/pybind/eager_method.cc [hunk 8]
```diff
@@ -152,6 +157,72 @@ static PyObject* eager_tensor_retain_grads(EagerTensorObject* self,
   EAGER_CATCH_AND_THROW_RETURN_NULL
 }
 
+static PyObject* eager_tensor__clear_gradient(EagerTensorObject* self,
+                                              PyObject* args,
+                                              PyObject* kwargs) {
+  EAGER_TRY
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_method.cc]: sync_try maybe

File: paddle/fluid/pybind/eager_method.cc [hunk 9]
```diff
@@ -152,6 +157,72 @@ static PyObject* eager_tensor_retain_grads(EagerTensorObject* self,
...
+  } else {
+    auto meta = egr::EagerUtils::unsafe_autograd_meta(self->eager_tensor);
+    grad = meta->Grad();
+  }
+
+  if (grad.initialized()) {
+    auto dense_tensor =
+        std::dynamic_pointer_cast<pten::DenseTensor>(grad.impl());
+    dense_tensor->release();
+  }
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_method.cc:194]: Add VLOG here to show grad is not initialized

File: python/paddle/fluid/dygraph/varbase_patch_methods.py [hunk 10]
```diff
@@ -280,15 +305,22 @@ def gradient(self):
                 # [500.]
 
         """
-        if self._grad_ivar() is None:
-            return None
+        if _in_eager_mode():
+            if self.grad is None:
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/varbase_patch_methods.py]: if self.grad._initialized()

Comment (JiabinYang) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: because there is no way to get `None` here

**Review by lanxianghit — DISMISSED** [review]
LGTM for deleting paddle.fluid.eager

**Review by JiabinYang — COMMENTED** [review]
some comment

**Review by zhiqiu — DISMISSED** [review]
LGTM for framework.py

**Review by TCChenlong — DISMISSED** [review]

File: paddle/fluid/pybind/eager_method.cc (hunk 7 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_method.cc]: done, thx!

File: paddle/fluid/pybind/eager_method.cc (hunk 8 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_method.cc]: done, thx!

File: paddle/fluid/pybind/eager_method.cc (hunk 9 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_method.cc:194]: done, thx!

File: python/paddle/fluid/dygraph/varbase_patch_methods.py (hunk 10 above)
Comment (wanghuancoder) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: done, thx!

**Review by JiabinYang — DISMISSED** [review]
LGTM

**Review by TCChenlong — DISMISSED** [review]

**Review by lanxianghit — DISMISSED** [review]

**Review by JiabinYang — COMMENTED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Pull request merged**